### PR TITLE
Add ResourceExplorer

### PR DESCRIPTION
**Pull request by zidaneymar**
Fixes #

## Description

1. Add resourceExplorer
2. Adapt typeloader test and declaritive sample bot.

## Specific Changes

  - 
  -
  -

### CODE REVIEW

File: libraries/botbuilder-dialogs-adaptive/src/actions/sendList.ts [hunk 1]
```diff
@@ -45,7 +45,7 @@ export class SendList extends Dialog {
     }
 
     protected onComputeId(): string {
-        return `SendList[]`;
+        return `SendList[${this.listProperty}]`;
```
Comment (luhan2017) [inline libraries/botbuilder-dialogs-adaptive/src/actions/sendList.ts]: Should we remove this action? C# don't have this, this functionality can be done through LG.

Comment (Stevenic) [reply libraries/botbuilder-dialogs-adaptive/src/actions/sendList.ts]: We can remove it assuming the LG pieces are working.

---
In reply to: [354679140](https://github.com/microsoft/botbuilder-js/pull/1476#discussion_r354679140) [](ancestors = 354679140)

File: libraries/botbuilder-dialogs-declarative/src/pathUtil.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
+import { normalize, join, resolve } from "path";
```
Comment (Stevenic) [inline libraries/botbuilder-dialogs-declarative/src/pathUtil.ts]: Needs header block

File: libraries/botbuilder-dialogs-declarative/src/pathUtil.ts [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+import { readdirSync, statSync, lstatSync, Stats } from "fs";
+
+export class PathUtil {
+    public static IsDirectory = source => lstatSync(source).isDirectory()
+    public static GetDirectories = source => readdirSync(source).map(name => join(source, name)).filter(PathUtil.IsDirectory)
+
+    public static GetAllFiles(dir: string): string[] {
```
Comment (Stevenic) [inline libraries/botbuilder-dialogs-declarative/src/pathUtil.ts]: >GetAllFiles [](start = 18, length = 11)

Method name case is wrong... Should be cammelCase

File: libraries/botbuilder-dialogs-declarative/src/resources/folderResoureProvider.ts [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+    private _emitter: EventEmitter;
+
+    set emitter(e: EventEmitter) {
+        this._emitter = e;
+    }
+
+    private resources: Map<string, FileResource> = new Map<string, FileResource>();
+
+    public IncludeSubFolders: boolean;
```
Comment (Stevenic) [inline libraries/botbuilder-dialogs-declarative/src/resources/folderResoureProvider.ts]: >IncludeSubFolders [](start = 11, length = 17)

Should be cammelCase.

File: libraries/botbuilder-dialogs-declarative/src/resources/resourceExplorer.ts [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+import { IResource } from "./resource";
+import { EventEmitter } from "events";
+
+export class ResourceExplorer {
+
+    private _emitter: EventEmitter = new EventEmitter();
+    private _resourceProviders: IResourceProvider[] = [];
+
+    constructor() {
+    }
```
Comment (Stevenic) [inline libraries/botbuilder-dialogs-declarative/src/resources/resourceExplorer.ts]: shouldn't need this

File: libraries/botbuilder-dialogs-declarative/src/resources/resourceExplorer.ts [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+    public async getResource(id: string): Promise<IResource> {
+        for (const rp of this._resourceProviders) {
+            const rpResource: IResource = await rp.getResource(id);
+            if (rpResource) {
+                return rpResource;
+            }
+        }
+
+        return null;
```
Comment (Stevenic) [inline libraries/botbuilder-dialogs-declarative/src/resources/resourceExplorer.ts]: >null [](start = 15, length = 4)

should be "undefined". avoid the use of "null"

**Pull request merged**